### PR TITLE
Report group by and calculated fields

### PR DESCRIPTION
**Pull request by mqueme**
…nd aggregated functions to reports

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR adds custom contact fields to all data sources that have contact fields, it adds group by options to reports, it adds aggregated function to reports
#### Steps to test this PR:
Apply PR, test that reports work, and that you can add group by and aggregated fields to  your reports, also see all contact column in you data sources that use contact fields
1. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

**Comment by escopecz** [comment]
I managed to break it on email data source with this configuration:
screen shot 2017-04-11 at 12 24 09
```
500 Internal Server Error - An exception occurred while executing 'SELECT cut.hits AS hits, CONCAT(ROUND(cut.hits/(e.sent_count * cut.trackable_count)*100),'%') AS hits_ratio, cut.unique_hits AS unique_hits, CONCAT(ROUND(cut.unique_hits/(e.sent_count * cut.trackable_count)*100),'%') AS unique_ratio, COUNT(DISTINCT dnc.id) AS unsubscribed, CONCAT(ROUND((COUNT(DISTINCT dnc.id)/e.sent_count)*100),'%') AS unsubscribed_ratio, e.sent_count AS sent_count, e.subject AS subject, CONCAT(ROUND((e.read_count/e.sent_count)*100),'%') AS read_ratio, e.read_count AS read_count, e.name AS e_name, COUNT() as 'COUNT' FROM emails e LEFT JOIN emails vp ON vp.id = e.variant_parent_id LEFT JOIN categories c ON c.id = e.category_id LEFT JOIN (SELECT COUNT(cut2.channel_id) AS trackable_count, SUM(cut2.hits) AS hits, SUM(cut2.unique_hits) AS unique_hits, cut2.channel_id FROM channel_url_trackables cut2 WHERE cut2.channel = 'email' GROUP BY cut2.channel_id) cut ON e.id = cut.channel_id LEFT JOIN lead_donotcontact dnc ON e.id = dnc.channel_id and dnc.channel='email' and dnc.reason=1 GROUP BY e.id, e.sent_count': SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ') as 'COUNT' FROM emails e LEFT JOIN emails vp ON vp.id = e.variant_parent_id LE' at line 1
```

**Comment by mqueme** [comment]
I have just tested same config. it worked for me. Did you clear your cache? regeerated assets?

**Comment by escopecz** [comment]
I found another:
screen shot 2017-04-12 at 12 16 52
```
500 Internal Server Error - An exception occurred while executing 'SELECT cut.hits AS hits, CONCAT(ROUND(cut.hits/(e.sent_count * cut.trackable_count)*100),'%') AS hits_ratio, cut.unique_hits AS unique_hits, CONCAT(ROUND(cut.unique_hits/(e.sent_count * cut.trackable_count)*100),'%') AS unique_ratio, COUNT(DISTINCT dnc.id) AS unsubscribed, CONCAT(ROUND((COUNT(DISTINCT dnc.id)/e.sent_count)*100),'%') AS unsubscribed_ratio, e.sent_count AS sent_count, e.subject AS subject, CONCAT(ROUND((e.read_count/e.sent_count)*100),'%') AS read_ratio, e.read_count AS read_count, e.name AS e_name, as 'COUNT' FROM emails e LEFT JOIN emails vp ON vp.id = e.variant_parent_id LEFT JOIN categories c ON c.id = e.category_id LEFT JOIN (SELECT COUNT(cut2.channel_id) AS trackable_count, SUM(cut2.hits) AS hits, SUM(cut2.unique_hits) AS unique_hits, cut2.channel_id FROM channel_url_trackables cut2 WHERE cut2.channel = 'email' GROUP BY cut2.channel_id) cut ON e.id = cut.channel_id LEFT JOIN lead_donotcontact dnc ON e.id = dnc.channel_id and dnc.channel='email' and dnc.reason=1 GROUP BY e.id, unique_ratio': SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'as 'COUNT' FROM emails e LEFT JOIN emails vp ON vp.id = e.variant_parent_id LEFT' at line 1
```

File: app/bundles/ReportBundle/Assets/js/report.js [hunk 1]
```diff
@@ -58,7 +59,7 @@ Mautic.addReportRow = function (elId) {
     prototypeHolder.append(output);
 
     var newColumnId = '#' + elId + '_' + index + '_column';
-
+    console.log(newColumnId);
```
Comment (escopecz) [inline app/bundles/ReportBundle/Assets/js/report.js]: If this debug line will be removed, I'll give this PR thumbs up :)

**Comment by mqueme** [comment]
We are changing this completely.

**Comment by mqueme** [comment]
removed and changes made

**Pull request merged**